mohawk/util.py

```diff
@@ -268,3 +275,110 @@ def normalize_header_attr(val):
     # Normalize like the hawk reference code.
     val = escape_header_attr(val)
     return val
+
+
+def get_bewit(resource):
+    """Returns a bewit identifier for the resource as a string"""
+    if resource.method != 'GET':
+        raise ValueError('bewits can only be generated for GET requests')
+    mac = calculate_mac(
+        'bewit',
+        resource,
+        None,
+    )
+
+    if isinstance(mac, six.binary_type):
+        mac = mac.decode('ascii')
+
+    if resource.ext is None:
+        ext = ''
+    else:
+        ext = resource.ext
+
+    # Strip out \ from the clientId
+    # since that can break parsing the response
+    # NB that the canonical implementation does not do this as of
+    # Oct 28, 2015, so this could break compat.
+    # We can leave \ in ext since validators can limit how many \ they split
+    # on (although again, the canonical implementation does not do this)
+    clientId = resource.credentials['id'].replace("\\", "")
+
+    # b64encode works only with bytes in python3, but all of our parameters are
+    # in unicode, so we need to encode them. The cleanest way to do this that
+    # works in both python 2 and 3 is to use string formatting to get a
+    # unicode string, and then explicitly encode it to bytes.
+    inner_bewit = u"{id}\\{exp}\\{mac}\\{ext}".format(
+        id=clientId,
+        exp=resource.timestamp,
+        mac=mac,
+        ext=ext,
+    )
+    inner_bewit_bytes = inner_bewit.encode('ascii')
+    bewit_bytes = urlsafe_b64encode(inner_bewit_bytes)
+    # Now decode the resulting bytes back to a unicode string
+    return bewit_bytes.decode('ascii')
+
+
+bewittuple = namedtuple('bewittuple', 'id exp mac ext')
+
+def parse_bewit(bewit):
+    """Returns the parts of the bewit as a namedtuple:
+        (id, exp, mac, ext)
+    Input it base64-encoded bewit string
+    """
+    decoded_bewit = b64decode(bewit).decode('ascii')
+    return bewittuple(*decoded_bewit.split("\\", 3))
+
+def strip_bewit(url):
+    """Strips the bewit parameter out of the url.
+    Returns (encoded_bewit, stripped_url)
+    Raises ValueError if no bewit found."""
+    m = re.search('[?&]bewit=([^&]+)', url)
+    if not m:
+        raise ValueError('no bewit found')
```

Comment:
I think this should also raise `InvalidBewit` (and a test is needed) because that way the authenticator could catch the root [HawkFail](https://github.com/catlee/mohawk/blob/bewit/mohawk/exc.py#L7) exception when responding to a request for a bewit protected URL.